File: prime_test/fermat_test/src/FermatTestDemo.kt
// Kotlin
// 素数判定: フェルマーテスト (Fermat Test)

import kotlin.random.Random

fun isPrime(target: Int, k: Int = 5): Boolean {
    if (target == 2 || target == 3) {
        return true
    }

    // 2未満または偶数は素数でない（2を除く）
    if (target < 2 || target % 2 == 0) {
        return false
    }

    repeat(k) {
        // 1 < a < target の範囲からランダムに数を選ぶ
        val a = Random.nextInt(2, target - 1)

        // フェルマーの小定理: a^(target-1) ≡ 1 (mod target) が成り立つか検証
        if (modPow(a.toBigInteger(), (target - 1).toBigInteger(), target.toBigInteger()) != 1.toBigInteger()) {
            return false
        }
    }

    return true
}

// modPow関数の実装（a^b mod m を計算）
fun modPow(base: java.math.BigInteger, exponent: java.math.BigInteger, modulus: java.math.BigInteger): java.math.BigInteger {
    return base.modPow(exponent, modulus)
}

fun main() {
    println("FermatTest TEST -----> start")

    println("\nisPrime")
    val inputList = listOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997)
    for (input in inputList) {
        val output = isPrime(input)
        println("  $input: $output")
    }

    println("\nFermatTest TEST <----- end")
}

File: prime_test/fermat_test/src/FermatTestDemo.cs
// C#
// 素数判定: フェルマーテスト (Fermat Test)

using System;
using System.Numerics;

class FermatTest
{
    static bool IsPrime(int target, int k = 5)
    {
        if (target == 2 || target == 3)
            return true;

        // 2未満または偶数は素数でない（2を除く）
        if (target < 2 || target % 2 == 0)
            return false;

        Random random = new Random();
        
        for (int i = 0; i < k; i++)
        {
            // 1 < a < target の範囲からランダムに数を選ぶ
            int a = random.Next(2, target - 1);

            // フェルマーの小定理: a^(target-1) ≡ 1 (mod target) が成り立つか検証
            // BigInteger.ModPow は (a^(target-1)) % target を効率的に計算します。
            if (BigInteger.ModPow(a, target - 1, target) != 1)
                return false;
        }

        return true;
    }

    static void Main()
    {
        Console.WriteLine("FermatTest TEST -----> start");

        Console.WriteLine("\nIsPrime");
        int[] inputList = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997 };
        foreach (int input in inputList)
        {
            bool output = IsPrime(input);
            Console.WriteLine($"  {input}: {output}");
        }

        Console.WriteLine("\nFermatTest TEST <----- end");
    }
}